File: src/main/scala/com/bytetrend/sandbox/scala/hackerrank/DiagonalDifference.scala
package com.bytetrend.sandbox.scala.hackerrank

import scala.collection.immutable.IndexedSeq

/**
  * Sample Input
  **
  *3
  *11 2 4
  *4 5 6
  *10 8 -12
  *Sample Output
  **
  *15
  *Explanation
  **
  *The primary diagonal is:
  *11
  *   5
  *     -12
  **
  *Sum across the primary diagonal: 11 + 5 - 12 = 4
  **
  *The secondary diagonal is:
  *      4
  *   5
  *10
  *Sum across the secondary diagonal: 4 + 5 + 10 = 19
  * Difference: |4 - 19| = 15
  */
object DiagonalDifference {

  /**
    * Receiving a list of tuples representing pairs from a matrix diagonals
    * Only values that are associated with a diagonal are non zero the main
    * diagonal values are in the first position of the pair the secondary
    * diagonal values are in the second position of the pair. We only need
    * to sum the first values with the first values and the same for the second
    * values.
    *
    * 0 = "(11,0)"
    * 1 = "(0,0)"
    * 2 = "(0,4)"
    * 3 = "(0,0)"
    * 4 = "(5,5)"
    * 5 = "(0,0)"
    * 6 = "(0,10)"
    * 7 = "(0,0)"
    * 8 = "(-12,0)"
    *
    * @param matrix
    * @return
    */
  def calc(matrix: Seq[(Int,Int)]): (Int,Int) ={
    matrix.reduceLeft((x,y) => (x._1+y._1,x._2+y._2))
  }

  def main(args:Array[String]): Unit ={
    val in  = new java.util.Scanner(System.in)
    val n= in.nextInt

    val matrix:IndexedSeq[(Int, Int)] = for {
      i <- 0 until n
      j <- 0 until n
    } yield{
      val z = in.nextInt
      //Primary diagonal is when the index i and j are the same we only capture those number in
      //the first place of a pair. The secondary diagonal is when the sum of i and j equal the n - 1
      //We only need to capture the diagonal row/column components all others can be zero.
      (if(i == j)z else 0,if(i + j == n - 1)z else 0)
    }
    val r = calc(matrix )
    println(math.abs(r._1-r._2))
  }
}

File: src/main/java/com/bytetrend/sandbox/java/hackerrank/DiagonalDifference.java
package com.bytetrend.sandbox.java.hackerrank;

import java.util.Scanner;

/**
 * Given a square matrix of size N x N , calculate the absolute difference between the sums of its diagonals.
 * <p>
 * Input Format
 * <p>
 * The first line contains a single integer, N. The next N lines denote the matrix's rows, with each line containing N space-separated integers describing the columns.
 * <p>
 * Output Format
 * <p>
 * Print the absolute difference between the two sums of the matrix's diagonals as a single integer.
 * <p>
 * Sample Input
 * <p>
 * 3
 * 11 2 4
 * 4 5 6
 * 10 8 -12
 * Sample Output
 * <p>
 * 15
 * Explanation
 * <p>
 * The primary diagonal is:
 * 11
 * 5
 * -12
 * <p>
 * Sum across the primary diagonal: 11 + 5 - 12 = 4
 * <p>
 * The secondary diagonal is:
 * 4
 * 5
 * 10
 * Sum across the secondary diagonal: 4 + 5 + 10 = 19
 * Difference: |4 - 19| = 15
 */
public class DiagonalDifference {
    //O(n^2)
    public static int calc1(int[][] array){
        int topRighDown = 0;
        int bottomRightUp = 0;

        for (int i = 0; i < array.length; i++) {
            for (int j = 0; j < array.length; j++) {
                if(i == j)
                    topRighDown += array[i][j];
                if(j + i == array.length - 1){
                    bottomRightUp += array[i][j];
                }
            }
        }
        return Math.abs(bottomRightUp-topRighDown);
    }
    //O(n)
    public static int calc2(int[][] array){
        int topRighDown = 0;
        int bottomRightUp = 0;
        for(int i =0; i< array.length; i++){
            topRighDown += array[i][i];
            bottomRightUp += array[i][(array.length-1)-i];
        }
        return Math.abs(bottomRightUp-topRighDown);
    }
    public static void main(String[] args) {

        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[][] array = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
               array[i][j] = in.nextInt();
            }
        }
        System.out.println(calc1(array));
        System.out.println(calc2(array));
    }
}


